samples/DevTestLabs/Modules/Library/CONTRIBUTING.md

```diff
@@ -0,0 +1,149 @@
+# Contributing
+
+When contributing to this repository, please first discuss the change you wish to make via issue,
+email, or any other method with the owners of this repository before making a change. 
+
+## Code Guidelines
+1. Put the `ValueFromPipeline` attribute on the parameter you expect to come from the pipeline.
+1. Return the correct object from your function (i.e. lab, vm or custom image) as to enable pipelining.
+2. Re-query for the resource you are returning from the function as your function might have changed it, so you likely need the latest one.
+3. Use the `begin/process/end` syntax and put `. BeginPreamble` in the `begin` part. This has to do with correct error management. Read the code for details.
+4. Use the `foreach` pattern to enable taking multiple objects from the pipeline
+5. Use the function below as a template on how to write the function
+
+```powershell
+function Invoke-AzDtlVmClaim {
+  [CmdletBinding()]
+  param(
+    [parameter(Mandatory=$true,HelpMessage="VM to stop. Noop if already stopped.", ValueFromPipeline=$true)]
+    [ValidateNotNullOrEmpty()]
+    $Vm
+  )
+
+  begin {. BeginPreamble}
+  process {
+    try {
+      foreach($v in $Vm) {
+        Invoke-AzureRmResourceAction -ResourceId $v.ResourceId -Action "claim" -Force | Out-Null
+        $v  | Get-AzDtlVm
+      }
+    } catch {
+      Write-Error -ErrorRecord $_ -EA $callerEA
+    }
+  }
+  end {}
+}
+```
+6. In the common case you are deploying an ARM template to a lab, use the utility function `DeployLab` as shown below
+
+```powershell
+function New-AzDtlLab {
+  [CmdletBinding()]
+  param(
+    [parameter(Mandatory=$true, ValueFromPipelineByPropertyName = $true, HelpMessage="Name of the lab to create")]
+    [ValidateLength(1,50)]
+    [ValidateNotNullOrEmpty()]
+    [string] $Name,
+
+    [parameter(Mandatory=$true, ValueFromPipelineByPropertyName = $true, HelpMessage="Name of the resource group to create the lab into. It must be an existing one.")]
+    [ValidateNotNullOrEmpty()]
+    [string] $ResourceGroupName,
+
+    [parameter(Mandatory=$false,HelpMessage="Run the command in a separate job")]
+    [switch] $AsJob = $False
+  )
+
+  begin {. BeginPreamble}
+  process {
+    try {
+      Get-AzureRmResourceGroup -Name $ResourceGroupName -ErrorAction Stop | Out-Null
```

Comment:
This is part of the whole error management strategy for the library as for previous comment. If we disagree on that, we need to change things more deeply (and it is a tricky business to get it right, so that it stops at first error inside the library code, but respects the error settings of the calling code).

Reply:
All good. As I went through the PR more I could see how we're intending it to be handled. Thank you.